jl4-lsp/src/LSP/L4/Viz/Ladder.hs

```diff
@@ -185,3 +225,21 @@ mkVizNameWith :: (Name -> Text) -> Resolved -> V.Name
 mkVizNameWith printer (getUniqueName -> (uniq, name)) =
   case uniq of
     MkUnique {unique} -> V.MkName unique (printer name)
+
+------------------------------------------------------
+-- Other helpers
+------------------------------------------------------
+
+exprHasBooleanType :: Expr Resolved -> Viz Bool
+exprHasBooleanType (getAnno -> Anno {extra = Extension {resolvedInfo = Just (TypeInfo ty _)}}) =
+  isBooleanType ty
+exprHasBooleanType _ = pure False
+
+-- | Returns True iff the Type Resolved is that of a L4 BOOLEAN
+isBooleanType :: Type' Resolved -> Viz Bool
+isBooleanType ty = do
+  type' <- getExpandedType ty
+  pure $ case type' of
+    TyApp _ (Ref _ uniq _) [] ->
+      uniq == TC.booleanUnique
+    _ -> False
```

Comment:
all of these are "pure" functions, they don't change the state. Normally you'd not put the environment in the state to avoid things like this. I think it's fine here though

Reply:
Yeah, I thought about using a Reader, but it also felt like it might be overkill.